lib/class-wp-theme-json.php

```diff
@@ -846,31 +847,60 @@ function ( $carry, $element ) {
 	 *     background: value;
 	 *   }
 	 *
+	 *   p.has-value-gradient-background {
+	 *     background: value;
+	 *   }
+	 *
 	 * @param string $stylesheet Stylesheet to append new rules to.
 	 * @param array  $context Context to be processed.
 	 *
 	 * @return string The new stylesheet.
 	 */
-	private static function to_stylesheet( $stylesheet, $context ) {
+	private static function to_css_variables( $stylesheet, $context ) {
 		if ( empty( $context['selector'] ) ) {
-			return '';
+			return $stylesheet;
 		}
 
 		$declarations = array();
-		self::compute_style_properties( $declarations, $context );
 		self::compute_preset_vars( $declarations, $context );
 		self::compute_theme_vars( $declarations, $context );
 
-		// If there are no declarations at this point,
-		// it won't have any preset classes either,
-		// so bail out earlier.
-		if ( empty( $declarations ) ) {
-			return '';
-		}
-
 		// Attach the ruleset for style and custom properties.
 		$stylesheet .= self::to_ruleset( $context['selector'], $declarations );
 
+		return $stylesheet;
+	}
+
+	/**
+	 * Converts each context into a list of rulesets
+	 * containing the block styles to be appended to the stylesheet.
+	 *
+	 * See glossary at https://developer.mozilla.org/en-US/docs/Web/CSS/Syntax
+	 *
+	 * For each context this creates a new ruleset such as:
+	 *
+	 *   context-selector {
+	 *     style-property-one: value;
+	 *   }
+	 *
+	 * @param string $stylesheet Stylesheet to append new rules to.
+	 * @param array  $context Context to be processed.
+	 *
+	 * @return string The new stylesheet.
+	 */
+	private static function to_block_styles( $stylesheet, $context ) {
+		if ( empty( $context['selector'] ) ) {
+			return $stylesheet;
+		}
+
+		$declarations = array();
+		self::compute_style_properties( $declarations, $context );
+
+		if ( ! empty( $declarations ) ) {
+			// Attach the ruleset for style and custom properties.
+			$stylesheet .= self::to_ruleset( $context['selector'], $declarations );
+		}
+
 		// Attach the rulesets for the classes.
 		self::compute_preset_classes( $stylesheet, $context );
```

Comment:
Conceptually, I don't like that we output the preset classes here. In practical terms, I guess it's fine? Flagging in case you have thoughts.